File: dynamic-programming-js/fib.js
const fib = (n) => {
  if (n <= 2) {
    return 1;
  }
  return fib(n - 1) + fib(n - 2);
};

// Memoize version
const fib2 = (n, memo={}) => {
  if (n in memo) return memo[n];
  if(n<=2) return 1;
  memo[n] = fib2(n - 1, memo) + fib(n - 2, memo);
  return memo[n]
}
//
// console.log(fib(6));
// console.log(fib(7));
// console.log(fib(50));


console.log(fib2(6));
console.log(fib2(7));
console.log(fib2(50));

File: AlgoExpert/Fib.py
def getNthFib(n):
    lastTwo = [0,1]
    counter = 3
    while counter <= n:
        nextFib = lastTwo[0] + lastTwo[1]
        lastTwo[0] = lastTwo[1]
        lastTwo[1] = nextFib
        counter += 1
    return lastTwo[1] if n > 1 else lastTwo[0]

'''
Time complexity: O(n)
Space complexity: O(1)
'''

##### Memoized version #######
def getNthFib(n, memo={1:0, 2:1}):
    if n in memo:
        return memo[n]
    else:
        memo[n] = getNthFib(n-1, memo) + getNthFib(n-2, memo)
    return memo[n]
